File: 2020/2020W37/preppindataw37.R
library(dplyr)
library(readxl)
library(tidyr)
library(lubridate)
library(bizdays)
library(fuzzyjoin)

Sys.setlocale("LC_ALL","English")

my_hol <- read_excel("F:/Data/Start Date.xlsx", sheet = "Holidays") %>%
  summarise('Holidays'= sum(`Holidays`)) %>%
  .[[1]]
tw_hol <-read_excel("F:/Data/Taiwan Holidays.xlsx", col_types = c("numeric", "text", "skip", "skip"))
start_date <- as.Date("2019/1/1")
to_date <- as.Date("2020/9/11")

tw_hol_list <- tw_hol %>%
  filter(grepl('.*to.*',`Date`)) %>%
  separate(`Date`,c('From Date', 'To Date'), sep=' to ') %>%
  mutate('From Date' = as.Date(paste(`From Date`,`Year`), format='%d %b %Y'),
         'To Date' = as.Date(paste(`To Date`,`Year`), format='%d %b %Y'),
         'Days' = time_length(`To Date` - `From Date`,unit='days'),
         'Max Days' = max(`Days`)->>maxseq) %>%
  fuzzy_inner_join(.,data.frame('Seq' =seq(0,maxseq)),
                   by = c('Days'='Seq'),
                   match_fun = list(`>=`)) %>%
  mutate('Date' = `From Date`  %m+% days(`Seq`)) %>% 
  select('Date') %>%
  rbind(.,tw_hol %>%
          filter(!grepl('.*to.*',`Date`)) %>%
          mutate('Date' = as.Date(as.numeric(`Date`),origin = "1899-12-30"),
                 'Excel Conv Year' = year(`Date`),
                 'Date' = `Date` %m+% years(`Year`-`Excel Conv Year`)) %>% 
          select('Date'))
 
cal <- create.calendar('mycal', holidays=pull(tw_hol_list, `Date`), 
                       weekdays=c("saturday", "sunday"),
                       start.date = start_date, end.date = to_date, financial=FALSE)

final <- data.frame('Start Date'=start_date, 'Today'=to_date,
                   'Working Days'= bizdays(start_date %m+% days(1),to_date,cal)-my_hol)

View(final)

File: 2020/2020W37/PreppindataW37.py
import pandas as pd
import numpy as np
from datetime import date,datetime,timedelta
import re

def gen_datelist(sd, ed):
    return [d.date() for d in pd.date_range(sd, periods=(ed-sd).days+1).tolist()]

myhol = pd.read_excel(pd.ExcelFile("F:/Data/Start Date.xlsx"),'Holidays')  
twhol = pd.read_excel(pd.ExcelFile("F:/Data/Taiwan Holidays.xlsx"),sheet_name=0)  
start_date = date(2019,1,1)
to_date = date(2020,9,11)

twhol['Is Date'] = twhol['Date'].apply(lambda x: 'Y' if isinstance(x, datetime) else 'N')

twgolrg = twhol[twhol['Is Date']=='N'].copy()
twgolrg['From date'] = twgolrg.apply(lambda x: datetime.strptime(
    re.search('(.*)(?=\sto)',x['Date']).group(1)+' '+str(x['Year']),'%d %b %Y').date(), axis=1) 
twgolrg['To date'] = twgolrg.apply(lambda x: datetime.strptime(
    re.search('(?<=to\s)(.*)',x['Date']).group(1)+' '+str(x['Year']),'%d %b %Y').date(), axis=1) 

twhol = twhol[twhol['Is Date']=='Y'].copy()
twhol['Date'] = twhol.apply(lambda x: x['Date'].replace(year=x['Year']).date(), axis=1)

twhol_list = twhol['Date'].tolist()
for _,row in twgolrg.iterrows():
    twhol_list = twhol_list + gen_datelist(row['From date'], row['To date'])  
    
final = [start_date,to_date,
         np.busday_count(start_date+timedelta(days=1), to_date+timedelta(days=1), holidays=twhol_list)-myhol['Holidays'].sum()]    
final = pd.DataFrame(final).T 
final.columns=['Start Date', 'Today', 'Working Days']
